### PR TITLE
feat(DatePicker): Add typescript type to DatePicker Component

### LINKED ISSUE #12540: Add TypeScript types to DatePicker, DatePickerInput

### PR DESCRIPTION
**Pull request by Shankar-CodeJunkie**
Closes #12540

##### Changed:
- DatePicker.js has been renamed to DatePicker.tsx
- Added interface and types on all the necessary places. 

##### Storybook testing

- I ran fresh yarn install, yarn build
- I moved into packages/react and ran `yarn storybook`
- I see the storybook opened on my localhost:3000 and I tested the DatePicker components and I don't see any failures on the screen.

### CODE REVIEW

**Comment by andreancardona** [comment]
@Shankar-CodeJunkie looks like you might need to run `yarn format` :)

**Comment by Shankar-CodeJunkie** [comment]
Thanks @andreancardona . I ran yarn format and pushed the changes to the branch. 

Let me correct the test failure as well

**Review by andreancardona — DISMISSED** [review]
@Shankar-CodeJunkie Looks good 🙌🏽

File: packages/react/src/components/DatePicker/DatePicker-test.js [hunk 1]
```diff
@@ -311,7 +311,9 @@ describe('Single date picker', () => {
     );
 
     userEvent.click(screen.getByText('clear'));
-    expect(screen.getByLabelText('Date Picker label')).toHaveValue('');
+    expect(screen.getByLabelText('Date Picker label')).toHaveValue(
```
Comment (tw15egan) [inline packages/react/src/components/DatePicker/DatePicker-test.js]: If we are testing that the value should clear when the button is pressed, this should test that clearing the selection works as intended. What has changed that is causing this test to fail now?

**Review by tw15egan — DISMISSED** [review]

File: packages/react/src/components/DatePicker/DatePicker-test.js (hunk 1 above)
Comment (Shankar-CodeJunkie) [reply packages/react/src/components/DatePicker/DatePicker-test.js]: @tw15egan .  Yes ! .. You are correct.  I'll check this and do the necessary changes and update it .

**Comment by Shankar-CodeJunkie** [comment]
@andreancardona : I did push some changes, but some of the checks are in pending state. Is there a way to re-trigger those checks

**Comment by Shankar-CodeJunkie** [comment]
@andreancardona - I see some test failures outside of the DatePicket component. 

```
 FAIL  packages/react/src/components/Copy/Copy-test.js
  ● Test suite failed to run

    TypeError: _userEvent.default.setup is not a function

      14 | jest.useFakeTimers();
      15 |
    > 16 | const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
         |                        ^
      17 |
      18 | describe('Copy', () => {
      19 |   it('should set tabIndex if one is passed via props', () => {

      at Object.setup (packages/react/src/components/Copy/Copy-test.js:16:24)

 FAIL  packages/react/src/components/CopyButton/CopyButton-test.js
  ● Test suite failed to run

    TypeError: _userEvent.default.setup is not a function

      12 |
      13 | jest.useFakeTimers();
    > 14 | const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
         |                        ^
      15 |
      16 | describe('CopyButton', () => {
      17 |   it('should set tabIndex if one is passed via props', () => {

      at Object.setup (packages/react/src/components/CopyButton/CopyButton-test.js:14:24)

 FAIL  packages/react/src/internal/__tests__/useDelayedState-test.js
  ● Test suite failed to run

    TypeError: _userEvent.default.setup is not a function

      12 |
      13 | jest.useFakeTimers();
    > 14 | const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
         |                        ^
      15 |
      16 | describe('useDelayedState', () => {
      17 |   it('should update the state after the given delayMs', async () => {

      at Object.setup (packages/react/src/internal/__tests__/useDelayedState-test.js:14:24)
```

File: packages/react/src/components/DatePicker/DatePicker.tsx [hunk 2]
```diff
@@ -183,6 +187,214 @@ function updateClassNames(calendar, prefix) {
...
+interface DatePickerProps
+  extends Omit<ReactAttr<HTMLDivElement>, ExcludedAttributes> {
+  /**
+   * flatpickr prop passthrough. Allows the user to enter a date directly
+   * into the input field
+   */
+  allowInput: boolean | undefined;
+
+  /**
+   * The DOM element the Flatpicker should be inserted into. `<body>` by default.
```
Comment (andreancardona) [inline packages/react/src/components/DatePicker/DatePicker.tsx]:
```suggestion
   * The DOM element the Flatpicker should be inserted into `<body>` by default.
```

File: packages/react/src/components/DatePicker/DatePicker.tsx [hunk 3]
```diff
@@ -183,6 +187,214 @@ function updateClassNames(calendar, prefix) {
...
+interface DatePickerProps
+  extends Omit<ReactAttr<HTMLDivElement>, ExcludedAttributes> {
+  /**
+   * flatpickr prop passthrough. Allows the user to enter a date directly
+   * into the input field
+   */
+  allowInput: boolean | undefined;
+
+  /**
+   * The DOM element the Flatpicker should be inserted into `<body>` by default.
```
Comment (andreancardona) [inline packages/react/src/components/DatePicker/DatePicker.tsx]:
```suggestion
   * The DOM element the flatpickr should be inserted into `<body>` by default.
```

File: packages/react/src/components/DatePicker/DatePicker-test.js (hunk 1 above)
Comment (Shankar-CodeJunkie) [reply packages/react/src/components/DatePicker/DatePicker-test.js]: @tw15egan : I have removed the changes on the DatePicket-test.js and made only this change, to get the test working.  
```
const [date, setDate] = useState('');
const [date, setDate] = useState();
```

File: packages/react/src/components/DatePicker/DatePicker.tsx [hunk 4]
```diff
@@ -27,6 +29,8 @@ import { usePrefix } from '../../internal/usePrefix';
 import { useSavedCallback } from '../../internal/useSavedCallback';
 import { FormContext } from '../FluidForm';
 import { WarningFilled, WarningAltFilled } from '@carbon/icons-react';
+import { ReactAttr } from '../../types/common';
+//import {FlatpickrFn} from "flatpickr/dist/types/instance";
```
Comment (tw15egan) [inline packages/react/src/components/DatePicker/DatePicker.tsx]: Can we delete this?

**Review by tw15egan — APPROVED** [review]
LGTM, just one comment 👍 ✅

File: packages/react/src/components/DatePicker/DatePicker.tsx (hunk 4 above)
Comment (Shankar-CodeJunkie) [reply packages/react/src/components/DatePicker/DatePicker.tsx]: Thanks @tw15egan . I've removed the line.

**Review by alisonjoseph — APPROVED** [review]

**Pull request merged**